### DISCUSSION TITLE
Variability in response times in a mixed cpu+io-bound application. Any way to improve ?

### DISCUSSION THREAD

**Discussion by mdespriee** [fastapi/fastapi#8344] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
None
```

### Description

We have a data intensive application exposed through FastAPI. Most endpoints do a mix of io-bound and cpu-bound work.
The whole app is using async clients to the datastores (asyncpg, aioredis), so it's coded in async all the way up to the endpoints.
It's deployed in bunch of containers behind a load-balancer (AWS ECS), so, following [fastapi docs](https://fastapi.tiangolo.com/deployment/docker/#replication-number-of-processes), we opted for uvicorn with 1 worker per container.

We notice a quite high variability in response time, and especially what seems to be "waiting times". 

When the response time is short, the profile looks like this:
![image](https://user-images.githubusercontent.com/1590580/178794318-784bb817-29aa-450b-8d1e-b08157d49658.png)
That is, the overall response time is a few ms more than the query time, so it's all good.

For some other calls, the profile looks like this:
![image](https://user-images.githubusercontent.com/1590580/178794711-22cbcc15-e916-4e1f-a4d6-97b1188ce2a5.png)
The overall response time is much longer than the longuest query, there are much more waiting time, especially at the beginning of the call, where there's essentially no work at all (any cpu-bound work comes after the sql response).

(Note: I'm just picking 2 examples on a fairly simple endpoint, but we have more complex endpoints with many DB calls, and the same pattern emerges, with waiting time popping here and there)

My understanding so far is that what looks like waiting time is just work being done on other concurrent requests. And we're just hitting the limits of the GIL-based archi of python and async behavior.

Is this understanding correct ?

From other issues I read, there may be some things to test, but I'm unsure:
- changing to gunicorn with n workers, to reduce the concurrency within the process ?
- isolate the cpu-bound work from the main thread. What would be the better ? wrapping the identified bottlenecks in run_in_threadpool ? or outright force the use of another thread at endpoint start ?

Anything else ?
Are there any recommendation to improve the situation ?

### Operating System

Linux

### FastAPI Version

0.78.0

### Python Version

3.9.13

**Comment by JarroVGIT** [comment]
> My understanding so far is that what looks like waiting time is just work being done on other concurrent requests. And we're just hitting the limits of the GIL-based archi of python and async behavior.
> 
> Is this understanding correct ?

It very well could be, but it is impossible to firmly confirm nor deny without all the data. 

>- changing to gunicorn with n workers, to reduce the concurrency within the process ?
>- isolate the cpu-bound work from the main thread. What would be the better ? wrapping the identified bottlenecks in run_in_threadpool ? or outright force the use of another thread at endpoint start ?

Running in K8S (which ECS basically is) means finetuning. There is not a single silver bullit, there are a few issues raised about this in this repo. I remember one guy doing some testing and he came to multiple workers per container, and multiple containers behind a LB for best performance. More workers (if they don't share the CPU cycles of course) would improve concurrency and reduce waiting time (if there is any right now!). 

Isolating your CPU bound work from the main thread would not be helpful I guess. Any task running in a thread is bound by the same event loop and GIL, so it wouldn't make a lot of difference. 

I'd guess you'd have to figure out if it really is waiting time because some other task is currently executing for another request. If that is the case, I would recommend starting with a bit more scale horizontally (e.g. more workers, either by increasing the amount of containers or increasing the amount of workers per container).

**Comment by iudeen** [comment]
We faced a very similar issue, which forced us to completely migrate to synchronous routes (to contain production issue) as we spent time on tuning our deployment to support async properly.

Here are somethings we did:

1. We used base docker image of python and ran as per the examples given in FastAPI docs. - one worker per container.

2. We had our load balancer and HPA set up in Kubernetes.

3. We implemented health check APIs and configured auto restart for our pods.

4. Made sure there's absolutely no blocking code in Python. Even made our logger asynchronous. 

5. Made sure we did not introduce memory leaks, connection leaks to DBs, increased session reusablity.

6. We made sure our middleware and dependency are async.

By performing above activities we were able to get predictable results with waiting time only increasing if all pods reach its maximum.

One issue we still face is, certain pods becoming unresponsive at times, which Health check API finds and restarts. 

Finally an understanding which I got (may be wrong): Async never means parallel rather it's concurrent. And it's not guaranteed that you won't have wait times. Wait times begin to surface when enough tasks are scheduled to run on a thread.
